### PR TITLE
Add neuron_from_equations(): generate a custom neuron class from state equations

### LINKED ISSUE #79: Enable custom neuron generators
Enable user to pass in state equations, to specify the state variable(s), and custom neuron class is automatically generated which can be treated like any other LIF neuron.

### PR DESCRIPTION
**Pull request by tritsystem**
# Add `neuron_from_equations()`: generate a custom neuron class from state equations

Closes #79.

## Motivation

> Enable user to pass in state equations, to specify the state
> variable(s), and custom neuron class is automatically generated which
> can be treated like any other LIF neuron.

Today, a model outside the built-in LIF family (Izhikevich, a custom
adaptive model, anything with more than one state variable) requires
hand-writing a `SpikingNeuron` subclass — see e.g. this session's own
`izhikevich` scratch branch, ~50 lines of boilerplate duplicating
`fire()`/reset bookkeeping that every other neuron already has for free.

## What this adds

```python
import snntorch as snn

def izh_update(states, I):
    v, u = states["v"], states["u"]
    dv = 0.04 * v ** 2 + 5 * v + 140 - u + I
    du = 0.02 * (0.2 * v - u)
    return {"v": v + dv, "u": u + du}

def izh_reset(states, spk, threshold):
    v, u = states["v"], states["u"]
    fired = spk.detach().bool()
    v = torch.where(fired, torch.full_like(v, -65.0), v)
    u = torch.where(fired, u + 8.0, u)
    return {"v": v, "u": u}

Izhikevich = snn.neuron_from_equations(
    "Izhikevich", state_names=("v", "u"), update_fn=izh_update,
    spike_state="v", reset_fn=izh_reset, init_values={"v": -65.0, "u": -13.0},
)

lif1 = Izhikevich(threshold=30.0)
spk, v, u = lif1(I)          # used exactly like any other neuron
```

`neuron_from_equations(name, state_names, update_fn, spike_state=None,
reset_fn=None, init_values=None)` returns a new class (a `CustomNeuron`
subclass, itself a `SpikingNeuron`). Every standard argument works:
`init_hidden`, `output`, `threshold`, `learn_threshold`, `spike_grad`,
`reset_mechanism`, `inhibition`, `state_quant`, `graded_spikes_factor`.

* **`update_fn(states, input_) -> states`** — the only required piece.
  Computes every state variable's next value from the current states and
  this step's input.
* **`spike_state`** — which state variable is compared to `threshold` to
  fire. Defaults to the first.
* **`reset_fn(states, spk, threshold) -> states`** — an escape hatch for
  models whose reset isn't subtract/zero/none of one state (Izhikevich's
  `v <- c, u <- u + d` touches *two* states on a spike). When omitted, the
  standard `reset_mechanism` is applied to `spike_state`, same as `Leaky`.

## Reset-timing note (documented, tested)

`CustomNeuron`'s built-in reset (when `reset_fn` is not given) is applied
in the **same step** as the firing decision — the common textbook
convention. This is exactly `snn.Leaky`'s behaviour with
`reset_delay=False`, **not** `Leaky`'s own default
(`reset_delay=True`, a deliberate one-step-delayed subtraction unique to
that class). Verified bit-for-bit: a generated neuron running `Leaky`'s
own update equation matches `snn.Leaky(beta=..., reset_delay=False)`
exactly over 20 random steps, batch 3×4.

## Evidence

| Check | Result |
|---|---|
| matches `Leaky(reset_delay=False)` | bit-for-bit, 20 steps, batch (3,4) |
| all 3 `reset_mechanism` options correct | subtract / zero / none each verified against a closed-form immediate-reset reference |
| genuinely different (2-state, custom reset) neuron | Izhikevich via the generator: 1–60 spikes over 300 steps at constant drive (bursts, doesn't saturate or go silent) — real dynamics, not a relabelled LIF |
| `init_hidden` / `output` | works, correct shapes |
| gradients | flow to input and `threshold` (`learn_threshold=True`) |
| validation | rejects missing `state_names`, bad `spike_state`, wrong state-tensor count, an `update_fn` that omits a state |

13 new tests. Full suite **223 passed / 2 xfailed** (was 210 / 2) — no
existing neuron file touched.

## Out of scope (follow-ups)

* A string/sympy-expression front-end over the Python-callable
  `update_fn` (issue says "pass in state equations" — a callable *is* the
  equation; a symbolic DSL is a much bigger, separate proposal).
* NIR export for generated neurons.

### CODE REVIEW

**Review by ixfd64 — CHANGES_REQUESTED** [review]
I see a few merge conflicts with your #453. Otherwise, the changes look good.

**Review by ixfd64 — APPROVED** [review]
